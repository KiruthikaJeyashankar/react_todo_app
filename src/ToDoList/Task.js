import './Task.css';

export default function Task({ task, tellParent }) {

    const toggleTask = () => {
        const newTask = { ...task, completed: !task.completed }
        console.log("newtask ",newTask)
        tellParent(newTask)
    }

    return <>
        {task.completed === true ?
            <li>
                <span onClick={() => toggleTask()}><s>{task.text}</s></span>
            </li>
            :
            <li>
                <span onClick={() => toggleTask()}>{task.text}</span>
            </li>
        }
    </>
};
