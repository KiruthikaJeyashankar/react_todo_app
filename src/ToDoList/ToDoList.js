import React from 'react';
import Task from './Task';
import { useState } from "react";

export default function ToDoList({ listOfTasks }) {

  const [todoList, setTodoList] = useState(listOfTasks);

  const handleParent = (updatedTask) => {
    setTodoList((currTodoList) =>
      currTodoList.map((task) =>
        task.id === updatedTask.id ? { ...task, completed: updatedTask.completed } : task
      )
    );
  }

  return (
    <div>
      {todoList.length === 0 && <h4><i>Nothing to do buddy. Sleep!!</i></h4>}
      <ul>
        {todoList.map((task) => <Task key={task.id} task={task} tellParent={(updatedTask) => handleParent(updatedTask)} />)}
      </ul>

      <br /> <br />
      <button data-testid="empty-list-button" type='button' onClick={() => setTodoList([])}>
        Empty List
      </button>

    </div>
  )
}
