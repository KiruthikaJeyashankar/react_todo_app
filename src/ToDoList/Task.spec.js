import React from 'react';
import { render } from '@testing-library/react';
import Task from './Task';
import '@testing-library/jest-dom/extend-expect';
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import ToDoList from "./ToDoList";

describe('Task Tests', () => {
    const task = {
        text: 'Test task',
        completed: false
    };

    it('should render task list', () => {
        const component = render(<Task task={task} tellParent={jest.fn()} />);

        expect(component.getByText(task.text)).toBeInTheDocument();
    });

    it('renders Task correctly', () => {
        const tree = renderer
            .create(<Task task={task} tellParent={jest.fn()} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should call tell parent when a task is toggled as completed', () => {
        let mockTellParent = jest.fn();

        const component = render(<Task task={task} tellParent={mockTellParent} />);

        expect(component.getByText(task.text)).toBeInTheDocument();

        userEvent.click(component.getByText(task.text));

        expect(mockTellParent).toBeCalledTimes(1);
        expect(mockTellParent).toBeCalledWith({ ...task, completed: true });
    });

    it('should call tell parent when a task is toggled as not completed', () => {
        const taskCompleted = {
            text: 'Test task',
            completed: true
        };
        let mockTellParent = jest.fn();

        const component = render(<Task task={taskCompleted} tellParent={mockTellParent} />);

        expect(component.getByText(taskCompleted.text)).toBeInTheDocument();

        userEvent.click(component.getByText(taskCompleted.text));

        expect(mockTellParent).toBeCalledTimes(1);
        expect(mockTellParent).toBeCalledWith({ ...task, completed: false });
    });

});
