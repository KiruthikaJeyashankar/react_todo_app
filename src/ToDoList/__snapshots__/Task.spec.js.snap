// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Tests renders TODOList correctly 1`] = `
<li>
  <span
    onClick={[Function]}
  >
    Test task
  </span>
</li>
`;
